src/hotspot/share/runtime/threadSMR.cpp

```diff
@@ -791,14 +791,19 @@ ThreadsListHandle::~ThreadsListHandle() {
 // associated ThreadsList. This ThreadsListHandle "protects" the
 // returned JavaThread *.
 //
+// If the thread reference is a mounted virtual thread and use_carrier is true,
+// return JavaThread for the carrier thread; if the thread is unmounted
+// or use_carrier is false, return false.
+//
```

Comment:
Right after the check on L827 below, what prevents the vthread from being unmounted?

So we fetch the JavaThread* on L833 and then verify that it is protected by the TLH
which is good, but that does not prevent the vthread from being unmounted by the
time that we return the JavaThread* on L853 below.

Am I missing something?

Reply:
Yes, that's right.
vthread can be unmounted at any moment. So it's caller responsibility to check is vthread is still mounted to the same carrier thread in handshake/VMOp (see comment about virtul threads in the threadSMR.hpp)
`VirtualThread.asyncGetStackTrace` retries to get stack trace in the case

Reply:
@dcubed-ojdk No you are not missing anything, the vthread could become unmounted. It is up to the subsequent handshake code to re-check that (as briefly discussed above).

Reply:
Except I'm now discovering that can't work other than to abandon the virtual thread in question if you found the carrier has changed, otherwise you are in a handshake with the wrong carrier and so can't grab the stack - but that means you need to retry for this virtual thread.

Reply:
I see the current code just produces an empty stack trace if the virtual thread is found to have switched carriers. This seems wrong. We should be "pinning" the virtual thread whilst we try to get its stacktrace, such that it is unmounted or else mounted and fixed to a carrier. Other thread walking code uses the JvmtiVTMSTransitionDisabler to try and achieve this.

Reply:
Sorry for being late to the party. I still need to walk through all the comments but it seems you have already caught the main problem with virtual threads. Nothing protects them from being concurrently mounted/unmounted/remounted. It is why the `JvmtiVTMSTransitionDisabler` is used in the JVMTI code. But, please, note that the `JvmtiVTMSTransitionDisabler` mechanism is enabled only when there is a JVMTI agent. Otherwise, it has been disabled for scalability purposes to exclude potentially high performance overhead at the VTMS transition points.

Reply:
If transition disabling is only available with JVMTI then I think we have a significant problem trying to deal with virtual threads in a number of APIs. Without disabling transitions you would basically need a loop that safely extracts the current carrier, handshakes with it, checks the VT is still mounted, and if not repeat the process. At the moment it appears that if the carrier is seen to have changed then we can just skip the VT - which would be very wrong IMO for most APIs e.g. thread dumps.

Reply:
> please, note that the JvmtiVTMSTransitionDisabler mechanism is enabled only when there is a JVMTI agent.

@sspitsyn  can you clarify this please. When you take a thread dump via the HotsotDiagnosticMXBean is there a JVMTI agent involved? Because that code uses the transition disabler.

Reply:
Thread::getStackTrace handles this with a SUSPEND bit and loop. I think the only non-JVMTI API where we've got an issue right now is HotsotDiagnosticMXBean.dumpThreads.

Reply:
> @sspitsyn can you clarify this please. When you take a thread dump via the HotsotDiagnosticMXBean is there a JVMTI agent involved? Because that code uses the transition disabler.

Yes, we have this issue for `HotsotDiagnosticMXBean.dumpThreads` as Alan has pointed out. We are aware about it and will evaluate a couple of fixing approaches.

Reply:
> Thread::getStackTrace handles this with a SUSPEND bit and loop.

Okay so that loop keeps retrying any time we get a null stacktrace from the VM, and unless we are extremely unlucky it should only take a couple of attempts to get an actual stack.

So that just leaves the issue of the carrier thread terminating, which is what this current fix was trying to address. So we need to ensure the carrier is protected by the TLH, which is what my other fix will address.

Reply:
BTW we need some comments making it clear that all null returns are handled via a retry at the higher level Java code.